src/ReflectionMethod.php

```diff
@@ -41,12 +49,82 @@ public function isStatic()
         return $this->method->isStatic();
     }
 
+    public function getReturnType()
+    {
+        if (!method_exists($this->method, 'getReturnType')) {
+            return null;
+        }
+
+        if ($type = $this->method->getReturnType()) {
+            // Self is not valid then used in different places.
+            if ('self' === $type->__toString()) {
+                return null;
+            }
+
+            return new ReflectionType($type);
+        }
+
+        return null;
+    }
+
     public function getDocComment()
     {
-        return preg_replace(
-            '/@(return|param) Generated\\\\/',
-            '@${1} ',
-            $this->method->getDocComment()
-        );
+        $pattern = '/@(return|param)\s+((?:[\w\\\\]+(?:\[\])?(?:\s*[|]\s*[\w\\\\]+(?:\[\])?)*))(\s|$)/';
+
+        return preg_replace_callback($pattern, [$this, 'processDocMatch'], $this->method->getDocComment());
+    }
+
+    private function processDocMatch(array $matches)
+    {
+        $types = [];
+        foreach (explode('|', $matches[2]) as $type) {
+            $types[] = $this->qualifyType(trim($type));
+        }
+
+        return sprintf('@%s %s%s', $matches[1], implode('|', $types), $matches[3]);
+    }
+
+    private function qualifyType($type)
+    {
+        // Type is fully qualified.
+        if ('\\' === $type[0]) {
+            return $type;
+        }
+
+        // Type is pointing to the Generated sub namespace.
+        if (0 === strpos($type, 'Generated\\')) {
+            return substr($type, 10);
+        }
+
+        // Handle array types
+        $array = '';
+        if ('[]' === substr($type, -2)) {
+            $type  = substr($type, 0, -2);
+            $array = '[]';
+        }
+
+        // Type can not be qualified.
+        if (in_array($type, ReflectionTypeInterface::NON_QUALIFIED_TYPES, true)) {
+            return $type . $array;
+        }
+
+        // Type resides in the namespace of the declaring class.
+        return sprintf('%s%s', $this->getResolveType($type), $array);
+    }
+
+    private function getResolveType($type)
+    {
+        $context_factory = new ContextFactory();
+        $context         = $context_factory->createFromReflector($this->method);
+        $fqn_resolver    = new FqsenResolver();
+
+        try {
+            $resolved_type = $fqn_resolver->resolve($type, $context);
+        } catch (\InvalidArgumentException $e) {
+            print_r($e);
+            return '>>>WHOOPS: ' . $type;
```

Comment:
Why catch this exception? Seems to me like a nice way to silently ignore the error, and make sure we forget the stack trace :wink: